### ISSUE TITLE
JavaNCSSCheck: unknown format type: numero inteiro

### ISSUE THREAD

**Issue by snepcadep** [checkstyle/checkstyle#2728]
Hi guys,

Please, could someone help me on this?

Working with Checkstyle plugin 2.4 on Sonar 5.2, we found something that looks like a bug. Apparently, one of our (big) Java classes presents some scenario that is not being correctly handled by Checkstyle. When Checkstyle runs over a specific Java class, it fails with the following exception:

```
ERROR: Error during Sonar runner execution
ERROR: Unable to execute Sonar
ERROR: Caused by: Can not execute Checkstyle
ERROR: Caused by: Exception was thrown while processing /path/to/class/FooBar.java
ERROR: Caused by: String index out of range: -1
```

Due to the complexity of the class, we weren't able to spot exactly what causes this error. Has anyone experienced something similar yet?

**Comment by romani** [comment]
Please report issue as http://checkstyle.sourceforge.net/report_issue.html#How_to_report_a_bug it will help to find a reason, we need to see full stacktrace i am not sure how to do this in sonar. Try to export sonar config to checkstyle config somehow and run our cli.

**Comment by romani** [comment]
If no details provided, issue will be closed in 3 weeks.

**Comment by snepcadep** [comment]
Here is the stacktrace when running the CLI:

```
java -jar checkstyle-6.12.1-all.jar -c checkstyle.xml ConsultasServiceImpl.java
Starting audit...
com.puppycrawl.tools.checkstyle.api.CheckstyleException: Exception was thrown while processing ConsultasServiceImpl.java
        at com.puppycrawl.tools.checkstyle.Checker.process(Checker.java:282)
        at com.puppycrawl.tools.checkstyle.Main.runCheckstyle(Main.java:276)
        at com.puppycrawl.tools.checkstyle.Main.main(Main.java:119)
Caused by: java.lang.IllegalArgumentException: unknown format type: numero inteiro
        at java.text.MessageFormat.makeFormat(Unknown Source)
        at java.text.MessageFormat.applyPattern(Unknown Source)
        at java.text.MessageFormat.<init>(Unknown Source)
        at com.puppycrawl.tools.checkstyle.api.LocalizedMessage.getMessage(LocalizedMessage.java:258)
        at com.puppycrawl.tools.checkstyle.api.LocalizedMessage.compareTo(LocalizedMessage.java:378)
        at com.puppycrawl.tools.checkstyle.api.LocalizedMessage.compareTo(LocalizedMessage.java:49)
        at java.util.TreeMap.put(Unknown Source)
        at java.util.TreeSet.add(Unknown Source)
        at com.puppycrawl.tools.checkstyle.api.LocalizedMessages.add(LocalizedMessages.java:53)
        at com.puppycrawl.tools.checkstyle.api.Check.log(Check.java:256)
        at com.puppycrawl.tools.checkstyle.checks.metrics.JavaNCSSCheck.leaveToken(JavaNCSSCheck.java:226)
        at com.puppycrawl.tools.checkstyle.TreeWalker.notifyLeave(TreeWalker.java:406)
        at com.puppycrawl.tools.checkstyle.TreeWalker.processIter(TreeWalker.java:504)
        at com.puppycrawl.tools.checkstyle.TreeWalker.walk(TreeWalker.java:332)
        at com.puppycrawl.tools.checkstyle.TreeWalker.processFiltered(TreeWalker.java:204)
        at com.puppycrawl.tools.checkstyle.api.AbstractFileSetCheck.process(AbstractFileSetCheck.java:79)
        at com.puppycrawl.tools.checkstyle.Checker.process(Checker.java:266)
        ... 2 more
Checkstyle ends with 1 errors.
```

**Comment by romani** [comment]
can you share with us checkstyle.xml and ConsultasServiceImpl.java ?

> numero inteiro

please share language/locale information of your OS.

**Comment by romani** [comment]
According to google tranlator your language is Portugies.

Problem is in inaccurately translated lines https://github.com/checkstyle/checkstyle/blob/master/src/main/resources/com/puppycrawl/tools/checkstyle/checks/metrics/messages_pt.properties#L28 . Text inside "{" should not be translated .

**Comment by snepcadep** [comment]
Hi Romani,

Yes, our language here is Portuguese.

We also found it strange, the `numero inteiro` message, but we haven't figured out why till you mention the locale. So, we executed the CLI again with English locale, and now we've got the original error again. Sorry about that.

```
java -jar checkstyle-6.12.1-all.jar -c checkstyle.xml ConsultasServiceImpl.java
Starting audit...
com.puppycrawl.tools.checkstyle.api.CheckstyleException: Exception was thrown while processing ConsultasServiceImpl.java
        at com.puppycrawl.tools.checkstyle.Checker.process(Checker.java:282)
        at com.puppycrawl.tools.checkstyle.Main.runCheckstyle(Main.java:276)
        at com.puppycrawl.tools.checkstyle.Main.main(Main.java:119)
Caused by: java.lang.StringIndexOutOfBoundsException: String index out of range: -1
        at java.lang.AbstractStringBuilder.charAt(Unknown Source)
        at java.lang.StringBuilder.charAt(Unknown Source)
        at com.puppycrawl.tools.checkstyle.checks.javadoc.JavadocStyleCheck.trimTail(JavadocStyleCheck.java:316)
        at com.puppycrawl.tools.checkstyle.checks.javadoc.JavadocStyleCheck.getCommentText(JavadocStyleCheck.java:269)
        at com.puppycrawl.tools.checkstyle.checks.javadoc.JavadocStyleCheck.checkFirstSentenceEnding(JavadocStyleCheck.java:230)
        at com.puppycrawl.tools.checkstyle.checks.javadoc.JavadocStyleCheck.checkComment(JavadocStyleCheck.java:207)
        at com.puppycrawl.tools.checkstyle.checks.javadoc.JavadocStyleCheck.visitToken(JavadocStyleCheck.java:148)
        at com.puppycrawl.tools.checkstyle.TreeWalker.notifyVisit(TreeWalker.java:390)
        at com.puppycrawl.tools.checkstyle.TreeWalker.processIter(TreeWalker.java:501)
        at com.puppycrawl.tools.checkstyle.TreeWalker.walk(TreeWalker.java:332)
        at com.puppycrawl.tools.checkstyle.TreeWalker.processFiltered(TreeWalker.java:204)
        at com.puppycrawl.tools.checkstyle.api.AbstractFileSetCheck.process(AbstractFileSetCheck.java:79)
        at com.puppycrawl.tools.checkstyle.Checker.process(Checker.java:266)
        ... 2 more
Checkstyle ends with 1 errors.
```

Here is "[checkstyle.xml](https://github.com/checkstyle/checkstyle/files/73869/checkstyle.txt)".

About "ConsultasServiceImpl.java", I apologize, but I'm afraid we won't be able to share it. I have to consult my team about this matter.

**Comment by snepcadep** [comment]
I analysed "messages_pt.properties" on line 28. Should it be:

```
ncss.method = NCSS para este método é {0, number, integer} (máximo permitido é {1, number, integer}).
```

Like lines 29 and 30?

Is this somehow related to the original error?

**Comment by romani** [comment]
> About "ConsultasServiceImpl.java", I apologize, but I'm afraid we won't be able to share it. I have to consult my team about this matter.

we do not need exact code, you can obfuscate it as much as you want. You can cut the file as much as you want . Change all types to `String` or anything else. Checkstyle is failing on some part of code not whole file. 

> Caused by: java.lang.StringIndexOutOfBoundsException: String index out of range: -1

This is another problem , not related to Locale. Problem in JavadocStyleCheck, so some javadoc in your file crash the it. Workaround is to disable this Check till issue is resolved. New issue need to be reported and config and input file.

**Comment by romani** [comment]
please create separate issue for JavadocStyleCheck

**Comment by rafquint** [comment]
I had the same problem. I am using Java 8, and maven-checkstyle-plugin:2.17.
To resolve this problem I just change to maven-checkstyle-plugin:2.16.
This worked , but would like to receive another tip.

**Comment by romani** [comment]
please recheck this page - https://maven.apache.org/plugins/maven-checkstyle-plugin/history.html
maven-checkstyle-plugin:2.17 using 6.11.2. This bug is fixed in 6.14.1.
Please update your plugin version 2.17 and use latest version of checkstyle library 6.16.1 - https://maven.apache.org/plugins/maven-checkstyle-plugin/examples/upgrading-checkstyle.html .

**Comment by rafquint** [comment]
Thanks Romani, it works!

<build>
    <pluginManagement>
        <plugins>
            <plugin>
                <groupId>org.apache.maven.plugins</groupId>
                <artifactId>maven-checkstyle-plugin</artifactId>
                <version>2.17</version>
                <dependencies>
                    <dependency>
                        <groupId>com.puppycrawl.tools</groupId>
                        <artifactId>checkstyle</artifactId>
                        <version>6.14.1</version>
                    </dependency>
                </dependencies>
            </plugin>
        </plugins>
    </pluginManagement>
</build>
